src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -486,15 +689,43 @@ - (void)startWithWriteable:(id<GRXWriteable>)writeable {
   // that the life of the instance is determined by this retain cycle.
   _retainSelf = self;
 
-  if (self.tokenProvider != nil) {
+  if (_callOptions == nil) {
+    GRPCMutableCallOptions *callOptions;
+    if ([GRPCHost isHostConfigured:_host]) {
+      GRPCHost *hostConfig = [GRPCHost hostWithAddress:_host];
+      callOptions = hostConfig.callOptions;
+    } else {
+      callOptions = [[GRPCMutableCallOptions alloc] init];
+    }
+    if (_serverName != nil) {
+      callOptions.serverAuthority = _serverName;
+    }
+    if (_timeout != 0) {
+      callOptions.timeout = _timeout;
+    }
+    uint32_t callFlags = [GRPCCall callFlagsForHost:_host path:_path];
+    if (callFlags != 0) {
+      if (callFlags == GRPC_INITIAL_METADATA_IDEMPOTENT_REQUEST) {
+        _callSafety = GRPCCallSafetyIdempotentRequest;
+      } else if (callFlags == GRPC_INITIAL_METADATA_CACHEABLE_REQUEST) {
+        _callSafety = GRPCCallSafetyCacheableRequest;
+      }
+    }
+
+    id<GRPCAuthorizationProtocol> tokenProvider = self.tokenProvider;
+    if (tokenProvider != nil) {
+      callOptions.authTokenProvider = tokenProvider;
+    }
+    _callOptions = callOptions;
+  }
+  if (_callOptions.authTokenProvider != nil) {
     self.isWaitingForToken = YES;
     __weak typeof(self) weakSelf = self;
```

Comment:
This is prexisting, but do we have a sense of why there's a strong/weak dance here? In general this API seems to be self-retaining around the duration of the call, which makes sense. Is there something particuar about auth tokens that might require breaking that pattern?

Reply:
I think when we introduced this, we followed the style of making a call weak in all the blocks. But I think you are right that making it strong should also work.